spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/audit/service/DefaultAuditRecordService.java

```diff
@@ -0,0 +1,87 @@
+ * limitations under the License.
+ */
+package org.springframework.cloud.dataflow.server.audit.service;
+
+import java.util.Map;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.springframework.cloud.dataflow.server.audit.domain.AuditActionType;
+import org.springframework.cloud.dataflow.server.audit.domain.AuditOperationType;
+import org.springframework.cloud.dataflow.server.audit.domain.AuditRecord;
+import org.springframework.cloud.dataflow.server.audit.repository.AuditRecordRepository;
+import org.springframework.util.Assert;
+
+
+/**
+ * Default implementation of the {@link AuditRecordService}.
+ *
+ * @author Gunnar Hillert
+ *
+ */
+public class DefaultAuditRecordService implements AuditRecordService {
+
+	private final AuditRecordRepository auditRecordRepository;
+
+	private final ObjectMapper objectMapper;
+
+	public DefaultAuditRecordService(AuditRecordRepository auditRecordRepository, ObjectMapper objectMapper) {
+		super();
+		Assert.notNull(auditRecordRepository, "auditRecordRepository must not be null.");
+		Assert.notNull(auditRecordRepository, "objectMapper must not be null.");
+		this.auditRecordRepository = auditRecordRepository;
+		this.objectMapper = objectMapper;
+	}
+
+	@Override
+	public void populateAndSaveAuditRecord(AuditActionType auditActionType, AuditOperationType auditOperationType,
+			String correlationId, String data) {
+		Assert.notNull(auditActionType, "auditActionType must not be null.");
+		Assert.notNull(auditOperationType, "auditOperationType must not be null.");
+		Assert.hasText(correlationId, "correlationId must not be null nor empty.");
+		Assert.hasText(data, "data must not be null nor empty.");
+
+		final AuditRecord auditRecord = new AuditRecord();
+		auditRecord.setAuditAction(auditActionType);;
+		auditRecord.setAuditOperation(auditOperationType);;
+		auditRecord.setCorrelationId(correlationId);
+		auditRecord.setAuditData(data);
+		this.auditRecordRepository.save(auditRecord);
+	}
+
+	@Override
+	public void populateAndSaveAuditRecord(AuditActionType auditActionType, AuditOperationType auditOperationType,
+			String correlationId, Map<String, Object> data) {
+		Assert.notNull(auditActionType, "auditActionType must not be null.");
+		Assert.notNull(auditOperationType, "auditOperationType must not be null.");
+		Assert.hasText(correlationId, "correlationId must not be null nor empty.");
+		Assert.notEmpty(data, "data map must not be null and must contain at least 1 entry.");
+
+		final AuditRecord auditRecord = new AuditRecord();
+		auditRecord.setAuditAction(auditActionType);;
+		auditRecord.setAuditOperation(auditOperationType);;
+		auditRecord.setCorrelationId(correlationId);
+
+		try {
+			auditRecord.setAuditData(objectMapper.writeValueAsString(data));
+		} catch (JsonProcessingException e) {
+			throw new IllegalStateException("Error serializing audit record data.", e);
```

Comment:
Should we proceed with the ongoing operation by logging the error when the audit save encounters error instead of throwing the exception here?

Reply:
Yea, we prob should write the record with the information as best we have it and some sort of entry for the data indicating it could not be saved.  Also some logging message.